src/System.Linq/src/System/Linq/Reverse.cs

```diff
@@ -20,7 +20,7 @@ public static IEnumerable<TSource> Reverse<TSource>(this IEnumerable<TSource> so
             return new ReverseIterator<TSource>(source);
         }
 
-        private sealed class ReverseIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
+        private sealed class ReverseIterator<TSource> : Iterator<TSource>, IPartition<TSource>
```

Comment:
I don't understand the value of this change / extra code.  This is simply to optimize the patterns:
```C#
source.Reverse().First();
source.Reverse().Last();
```
?  Why wouldn't the author instead just write the simpler and faster:
```C#
source.Last();
source.First();
```
?